org.lflang/src/org/lflang/validation/LFValidator.java

```diff
@@ -1688,11 +1713,162 @@ private boolean sameType(Type type1, Type type2) {
 
     private static String RESERVED_MESSAGE = "Reserved words in the target language are not allowed for objects "
             + "(inputs, outputs, actions, timers, parameters, state, reactor definitions, and reactor instantiation): ";
-    
+
     private static List<String> SPACING_VIOLATION_POLICIES = List.of("defer", "drop", "replace");
-    
+
     private static String UNDERSCORE_MESSAGE = "Names of objects (inputs, outputs, actions, timers, parameters, "
             + "state, reactor definitions, and reactor instantiation) may not start with \"__\": ";
-    
+
     private static String USERNAME_REGEX = "^[a-z_]([a-z0-9_-]{0,31}|[a-z0-9_-]{0,30}\\$)$";
+
+    /** A map from a string to a supported AttributeSpec */
+    private static final Map<String, AttributeSpec> ATTRIBUTE_SPECS_BY_NAME = new HashMap<>();
+
+    //////////////////////////////////////////////////////////////
+    //// Inner classes.
+    /**
+     * Specification of the structure of an annotation.
+     */
+    static class AttributeSpec {
```

Comment:
I think I raised this point before, but why aren't we putting this class in its own file?

Reply:
Just now read your reply https://github.com/lf-lang/lingua-franca/pull/977#discussion_r902077973 (let's continue there and resolve this comment.